### ISSUE TITLE
weird loss curve when finetuning with gigaspeech pretrained k2 model

### ISSUE THREAD

**Issue by nullscc** [k2-fsa/icefall#925]
My utimate goal is to compare wenet and k2 on online streaming situation for my team. To convince my team come to k2 for online streaming.

But I'm stuck on the first step: finetuning gigaspeech pretrained k2 model from https://huggingface.co/wgb14/icefall-asr-gigaspeech-pruned-transducer-stateless2(pretrained-iter-3488000-avg-15.pt) in offline decode mode, on about 2.4k hours data.

What I first did is:
1. `git lfs install && git clone https://huggingface.co/wgb14/icefall-asr-gigaspeech-pruned-transducer-stateless2`
2. `git clone https://github.com/k2-fsa/icefall && cd icefall/egs && cp -R gigaspeech myrecipe` 
3. `cd myrecipe`, modify `pruned_transducer_stateless2/asr_datamodule.py` to handle my dataset
4. `cd pruned_transducer_stateless2 && mkdir exp && cd exp && ln -s icefall-asr-gigaspeech-pruned-transducer-stateless2/exp/pretrained-iter-3488000-avg-15.pt epoch-1.pt`
5. use this command to train
  ```
./pruned_transducer_stateless2/train.py \
  --max-duration 400 \
  --num-workers 8 \
  --world-size 2 \
  --exp-dir pruned_transducer_stateless2/exp \
  --start-epoch 2 \
  --bpe-model data/lang_bpe_500/bpe.model \
  --on-the-fly-feats False \
  --enable-spec-aug True \
  --use-fp16 True
  ``` 

Then I get this tensorboard loss curve:
<img width="1420" alt="finetune_on_2 4k_lr_0 003" src="https://user-images.githubusercontent.com/17700282/221189043-7f1efe94-bbe3-43f1-8248-5191df5bdb50.png">

Noticed that weird loss curve, I adjust by these steps:
1. change `--initial-lr` to 0.0006
2. change `--lr-epochs` to 24

But I still get weird loss curve as blow:
<img width="1415" alt="finetune_on_2 4k_lr_0 0006" src="https://user-images.githubusercontent.com/17700282/221190059-a9f92720-15a7-48cf-b05d-44b050621c7b.png">

Still...

I change `--initial-lr` to 0.00009 and freeze encoder and decoder using following code:
```
    for t in ['encoder', 'decoder']:
        for k, p in model.named_parameters():
            if k.startswith(t + ".") or k == t:
                logging.info(f"Setting {k}.requires_grad = False")
                p.requires_grad = False
```
Then loss curve:
<img width="1413" alt="finetune_on_2 4k_lr_0 00009_joiner" src="https://user-images.githubusercontent.com/17700282/221190498-34bcbe25-8ee8-488e-aa41-bbfc5b76acad.png">

Still...

Then I use the same code to continue finetunning gigaspeech `S` subset using above modified code and strategy(still freeze encoder and decoder), and using following script to train:
```
./pruned_transducer_stateless2/train_modifi.py \
  --max-duration 400 \
  --num-workers 8 \
  --world-size 2 \
  --exp-dir pruned_transducer_stateless2/exp \
  --start-epoch 2 \
  --bpe-model data/lang_bpe_500/bpe.model \
  --on-the-fly-feats False \
  --enable-musan False \
  --subset S \
  --small-dev True \
  --enable-spec-aug True \
  --use-fp16 True
```
I still get:
<img width="1408" alt="finetune_on_gigaspeech_S" src="https://user-images.githubusercontent.com/17700282/221190795-a293705c-7901-45a8-9ac3-af39cee1acbe.png">

Version infomation:
* k2, builded from source using branch master, commit: a8e65478f2ba087da80a92fa261e21d6015f62aa
* lhotse, install from branch master and commit: a8e65478f2ba087da80a92fa261e21d6015f62aa by `pip install -e .`
* kaldifeat, builded from source using branch master, commit: 9fdc43df23e0a5c9d5b65b371c67d9f06807a402
* icefall, using branch master, commit: bffce413f07d938e35d69d5eb2f360c7ff842502
* pytorch, 1.11.0+cu113
* cuda, 11.3
* cudnn, 8.2.0
* torchaudio, 0.11.0+cu113

I noticed that sharp ascent on loss curve always happened on step 40k. Is this normal?

I really need help to handle this problem, any tips would be appreciated.

**Comment by marcoyang1998** [comment]
`Pruned_transducer_stateless2` is a rather old recipe. We’ve done a slight change to the optimizer in this PR   #550, where we clamp all the scalar parameters. I believe this commit is after the pretrained model, which might cause some mismatch during fine tuning . @csukuangfj What do you think?

**Comment by yaozengwei** [comment]
> `Pruned_transducer_stateless2` is a rather old recipe. We’ve done a slight change to the optimizer in this PR #550, where we clamp all the scalar parameters. I believe this commit is after the pretrained model, which might cause some mismatch during fine tuning . @csukuangfj What do you think?

I think it might be the reason. @nullscc Maybe you could firstly comment that change in `optim.py`, and try that again with a small --initial-lr,  to see whether it still happens. 
```python
# Constrain the range of scalar weights
if p.numel() == 1:
    p.clamp_(min=-10, max=2)
```

**Comment by danpovey** [comment]
I think the reason for the discontinuity at 40k batches is because of model_warm_step=20000.  At twice the warm-step, it changes the scale on the pruned loss.
Also the model behaves differently  below the warm-step.  You could step model_warm_step=1, that may fix the issue.

**Comment by danpovey** [comment]
Also, I suggest to check what the learning rate was in the final version of the model (it may be present in one of the fields of the .pt file,  use torch.load and poke around there).  You could start the learning rate at, say, twice that value and gradually decay it.  The Noam scheduler, which has a warmup period, may not be super-ideal.  Personally I would prefer a schedule without warmup, but it probably doesn't matter that much.

**Comment by wangtiance** [comment]
That recipe uses a "jumpy" pruned_loss_scale which causes the sudden increase in total loss. If you look at simple loss and pruned loss they are both decreasing. You can do what Dan suggested, or look at a more recent recipe like https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/pruned_transducer_stateless7_streaming and adapt it to gigaspeech dataset. It uses a smooth loss scale and supports streaming decode.

**Comment by nullscc** [comment]
> Also, I suggest to check what the learning rate was in the final version of the model (it may be present in one of the fields of the .pt file, use torch.load and poke around there). You could start the learning rate at, say, twice that value and gradually decay it. The Noam scheduler, which has a warmup period, may not be super-ideal. Personally I would prefer a schedule without warmup, but it probably doesn't matter that much.

Yes, I did find it previously in avg-15(avg-20 hasn't this value), do you mean if I want to finetune on a pretrained model, I need set learning rate twice as the stored value, so as to continue with a smooth learning rate and get a not so slow finetune process?

Thank you.

**Comment by danpovey** [comment]
Yes... using a larger learning rate will tend to make the model noisy and "undo" previously learned things, but will also help you to learn your data quickly.  You should decay the learning rate to at least as small to the final learning rate of the trained model.  You can also try starting from the final learning rate.  All our learning rates are on the high side.

**Comment by nullscc** [comment]
Thank you all, to keep just one thing changed, I first just tried `model_warm_step=1` in `train.py`showed as below:
<img width="1052" alt="image" src="https://user-images.githubusercontent.com/17700282/221224998-bf1cfe85-fece-40f0-82fc-c41d271f34c4.png">

the loss curve still jump on 40k step:
<img width="1409" alt="image" src="https://user-images.githubusercontent.com/17700282/221225349-9167879f-da70-4517-8187-2bd443a89df5.png">

**Comment by danpovey** [comment]
I suspect you made some mistake, e.g. viewed an old experiment directory, forgot to save the file, etc.

**Comment by nullscc** [comment]
> Yes... using a larger learning rate will tend to make the model noisy and "undo" previously learned things, but will also help you to learn your data quickly. You should decay the learning rate to at least as small to the final learning rate of the trained model. You can also try starting from the final learning rate. All our learning rates are on the high side. [Edit: it might also be a good idea to dump the averaged-over-time version of the model and try to start from that point, it will be less noisy. @yaozengwei knows how to do this, he added an option for doing this to one of the decode.py scripts.]

Thank you, it's a very import finetune trick for me.

**Comment by nullscc** [comment]
> 

Yes, I've noticed that so I recalled that comment~ 

And I double checked `model_warm_step` value by load saved .pt file, and the value is 1.

**Comment by nullscc** [comment]
Thank you all for such a huge help. 

I'll update the result tomorrow.

**Comment by yaozengwei** [comment]
The model that @nullscc has downloaded https://huggingface.co/wgb14/icefall-asr-gigaspeech-pruned-transducer-stateless2(pretrained-iter-3488000-avg-15.pt)  is an average model over some iterations. But we don't make implement the average over time (see https://github.com/k2-fsa/icefall/pull/344 and https://github.com/k2-fsa/icefall/issues/715#issuecomment-1330609308) in that recipe. Maybe @nullscc could refer to https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless4/train.py, to save a averaged model over time in the future experiments, which would give a better averaged model with fewer noise.

**Comment by nullscc** [comment]
> 

Yes, I've decode before. Blow was very first time I finetuned on my 2.4k hours data:
<img width="1416" alt="image" src="https://user-images.githubusercontent.com/17700282/221239780-d74e3987-40f8-4b40-9344-5dfc19d98450.png">

Decoding result on two of our test-set:
<html xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:x="urn:schemas-microsoft-com:office:excel" xmlns="http://www.w3.org/TR/REC-html40">
<head>

<meta name=Generator content="Microsoft Excel">


</head>
<body>



  | testset_1 | testset_2
-- | -- | --
epoch-3 | 52.14 | 32.3
epoch-9 | 42.08 | 26.12
epoch-12 | 47.51 | 29.79
epoch-19 | 38.59 | 25.54
epoch-28 | 24.41 | 24.41
wenet model finetune on gigaspeech | 17.68 | 15.26



</body>

</html>

I remember there was another sudden ascent on loss curve, so I suspect there maybe some problem in  my finetune process.(But can't find this kind of loss curve now, maybe just prejudice due to long time worse result)

Is this normal? And should I continue finetuning with above model to get a competable result? If so any suggestion to speed up the finetunning process~

**Comment by nullscc** [comment]
It seems there is no "jump" in loss curve when 40k step after I set `model_warm_step=1`
<img width="1411" alt="image" src="https://user-images.githubusercontent.com/17700282/221346469-786093a5-cd15-4ca8-a8af-59d6b0e1cb4e.png">
<img width="1406" alt="image" src="https://user-images.githubusercontent.com/17700282/221346492-9f77c116-8c54-4770-8045-e8c9b17fce6e.png">

But I still not sure if I could get a good finetune result~.

Any way, this "weird loss" "problem" seems solved. I'll close this issue.

And maybe open another issue to discuss how to get good finetune result.

Thank you all very much.

**Comment by csukuangfj** [comment]
@nullscc 
I suggest that you use changes from
https://github.com/k2-fsa/icefall/pull/344

(We have been using it since pruned_transducer_stateless4. You are using pruned_transducer_stateless2, which does not have this change)

**Comment by nullscc** [comment]
> @nullscc I suggest that you use changes from #344
> 
> (We have been using it since pruned_transducer_stateless4. You are using pruned_transducer_stateless2, which does not have this change)

Thank you for telling me that. From my view, compare to pruned_transducer_stateless2, pruned_transducer_stateless4 adds some model average  stratege and delay penalty, and it will slightly increase the performance.

But my ultimate goal is online streaming on k2(zipformer is best for this scene?). This finetuening on k2 gigaspeech pretrained model is just a evaluation step, and let us familiar with k2, it doesn't much matter slightly better performance with this strategy in offline decoding situation.

<img width="1005" alt="image" src="https://user-images.githubusercontent.com/17700282/221349220-ee652e91-acc0-4c84-b8ba-41f555cd483e.png">

<img width="1107" alt="image" src="https://user-images.githubusercontent.com/17700282/221349334-f6d5e057-13ac-4f3b-85d5-18b792dd976d.png">

So in this step, I just want to get a competable result compare to wenet. From now on, I have following finetuning result by directly load gigaspeech pretrained-iter-3488000-avg-15.pt, and finetuing with `/home/zhiwei001/icefall/egs/sgeng/ASR/pruned_transducer_stateless2/train.py` without any modification:
<html xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:x="urn:schemas-microsoft-com:office:excel" xmlns="http://www.w3.org/TR/REC-html40">
<head>

<meta name=Generator content="Microsoft Excel">


</head>
<body>



  | gigaspeech-test | testset2 | testset3 | testset4 | testset5
-- | -- | -- | -- | -- | --
wenet model finetune on gigaspeech | 17.68 | 15.26 | 9.8 | 11.6 | 8.51
epoch-0(from k2 gigaspeech pretrained model) | 10.52 | 44.02 | 68.02 | 53.67 | 23.21
epoch-3 | 52.14 | 32.3 | 21.7 | 25.68 | 14.8
epoch-9 | 42.08 | 26.12 | 18.29 | 20.88 | 12.81
epoch-12 | 47.51 | 29.79 | 19.89 | 23.09 | 14.09
epoch-19 | 38.59 | 25.54 | 17.79 | 20.18 | 12.37
epoch-28 | 39.22 | 24.41 | 17.29 | 19.36 | 12.22
epoch-32 | 37.4 | 23.56 | 16.85 | 19.14 | 12.14
epoch-35 | 39.03 | 23.6 | 16.96 | 19.36 | 12.19



</body>

</html>


I'm afraid I can't get a get a competable result compare to wenet~
